File: autolink/back/android/src/main/java/com/lynxapp/autolink/back/BackModule.java
package com.lynxapp.autolink.back;

import android.content.Context;
import android.content.ContextWrapper;
import android.os.Handler;
import android.os.Looper;

import androidx.activity.BackEventCompat;
import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;

import com.lynx.jsbridge.LynxContextModule;
import com.lynx.jsbridge.LynxMethod;
import com.lynx.jsbridge.LynxNativeModule;
import com.lynx.react.bridge.Callback;
import com.lynx.react.bridge.JavaOnlyArray;
import com.lynx.react.bridge.JavaOnlyMap;
import com.lynx.tasm.LynxView;
import com.lynx.tasm.behavior.LynxContext;

/** Lifecycle-bound Back dispatcher with an optional native animation target. */
@LynxNativeModule(name = BackModule.NAME)
public final class BackModule extends LynxContextModule {
    public static final String NAME = "Back";
    private static final String EVENT_NAME = "back";

    private static final String PLATFORM_ANDROID = "android";
    private static final String PHASE_START = "start";
    private static final String PHASE_PROGRESS = "progress";
    private static final String PHASE_CANCEL = "cancel";
    private static final String PHASE_COMMIT = "commit";
    private static final String SOURCE_SYSTEM = "system";
    private static final String SOURCE_GESTURE = "gesture";
    private static final String EDGE_LEFT = "left";
    private static final String EDGE_RIGHT = "right";
    private static final String EDGE_NONE = "none";

    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    @Nullable private final FragmentActivity activity;
    private boolean registered;
    private boolean destroyed;
    private boolean desiredEnabled;
    private long configuredRevision;
    @NonNull private String configuredInterceptorId = "";
    @NonNull private String configuredTargetId = "";

    private boolean gestureStarted;
    private boolean gestureFinishing;
    private long nextGestureId;
    private long gestureId;
    @NonNull private String gestureInterceptorId = "";
    @Nullable
    private PredictiveBackOverlayElement.PredictiveBackOverlayView gestureTarget;
    private double lastProgress;
    private double lastTouchX;
    private double lastTouchY;
    @NonNull private String lastEdge = EDGE_NONE;

    private final OnBackPressedCallback backCallback =
            new OnBackPressedCallback(false) {
                @Override
                public void handleOnBackStarted(@NonNull BackEventCompat event) {
                    if (gestureStarted || gestureFinishing) {
                        return;
                    }
                    updateGesture(event);
                    beginGesture(SOURCE_GESTURE);
                }

                @Override
                public void handleOnBackProgressed(@NonNull BackEventCompat event) {
                    if (gestureFinishing) {
                        return;
                    }
                    updateGesture(event);
                    if (!gestureStarted) {
                        beginGesture(SOURCE_GESTURE);
                    }
                    PredictiveBackOverlayElement.PredictiveBackOverlayView target =
                            gestureTarget;
                    if (target != null) {
                        target.update((float) lastProgress, lastEdge);
                    } else {
                        emitCurrent(PHASE_PROGRESS, SOURCE_GESTURE);
                    }
                }

                @Override
                public void handleOnBackCancelled() {
                    cancelGesture(true);
                }

                @Override
                public void handleOnBackPressed() {
                    if (gestureStarted) {
                        commitGesture(SOURCE_GESTURE);
                    } else {
                        emitDiscreteBack();
                    }
                }
            };

    public BackModule(LynxContext context) {
        super(context);
        activity = resolveFragmentActivity(context);
    }

    /** Compatibility switch for headless consumers that do not use backStack. */
    @LynxMethod
    public void setEnabled(boolean enabled, Callback callback) {
        mainHandler.post(() -> {
            configuredRevision += 1;
            configuredInterceptorId = "";
            configuredTargetId = "";
            desiredEnabled = enabled;
            if (!enabled) {
                cancelGesture(true);
            }
            applyDesiredInterception(callback);
        });
    }

    /** Atomically installs the complete top-of-stack snapshot from TypeScript. */
    @LynxMethod
    public void configure(
            boolean enabled,
            String interceptorId,
            String targetId,
            double revision,
            Callback callback) {
        mainHandler.post(() -> {
            if (destroyed) {
                callback.invoke("Back has already been destroyed");
                return;
            }
            long nextRevision = Math.max(0, (long) revision);
            if (nextRevision < configuredRevision) {
                callback.invoke("");
                return;
            }
            configuredRevision = nextRevision;
            desiredEnabled = enabled;
            configuredInterceptorId = interceptorId == null ? "" : interceptorId;
            configuredTargetId = targetId == null ? "" : targetId;
            // Keep the callback installed until the gesture that captured the
            // previous snapshot terminates. The next gesture gets this config.
            if (gestureStarted || gestureFinishing) {
                callback.invoke("");
                return;
            }
            applyDesiredInterception(callback);
        });
    }

    @Override
    public void destroy() {
        destroyed = true;
        mainHandler.post(() -> {
            cancelGesture(false);
            desiredEnabled = false;
            backCallback.setEnabled(false);
            backCallback.remove();
            registered = false;
        });
    }

    private void emitDiscreteBack() {
        lastProgress = 0;
        lastTouchX = 0;
        lastTouchY = 0;
        lastEdge = EDGE_NONE;
        beginGesture(SOURCE_SYSTEM);
        commitGesture(SOURCE_SYSTEM);
    }

    private void beginGesture(@NonNull String source) {
        gestureStarted = true;
        gestureFinishing = false;
        gestureId = ++nextGestureId;
        gestureInterceptorId = configuredInterceptorId;
        gestureTarget = PredictiveBackOverlayElement.findTarget(
                mLynxContext, configuredTargetId);
        PredictiveBackOverlayElement.PredictiveBackOverlayView target = gestureTarget;
        if (target != null) {
            target.begin(lastEdge);
            target.update((float) lastProgress, lastEdge);
        }
        emitCurrent(PHASE_START, source);
    }

    private void commitGesture(@NonNull String source) {
        if (!gestureStarted || gestureFinishing) {
            return;
        }
        PredictiveBackOverlayElement.PredictiveBackOverlayView target = gestureTarget;
        if (target == null) {
            emit(PHASE_COMMIT, 1, source, lastEdge, lastTouchX, lastTouchY);
            resetGesture();
            applyDesiredInterception(null);
            return;
        }

        gestureFinishing = true;
        long finishingGestureId = gestureId;
        target.commit(() -> {
            if (destroyed || !gestureStarted || gestureId != finishingGestureId) {
                return;
            }
            emit(PHASE_COMMIT, 1, source, lastEdge, lastTouchX, lastTouchY);
            resetGesture();
            applyDesiredInterception(null);
        });
    }

    private void cancelGesture(boolean emitEvent) {
        if (!gestureStarted) {
            return;
        }
        PredictiveBackOverlayElement.PredictiveBackOverlayView target = gestureTarget;
        if (target != null) {
            target.cancel();
        }
        if (emitEvent) {
            emitCurrent(PHASE_CANCEL, SOURCE_GESTURE);
        }
        resetGesture();
        if (!destroyed) {
            applyDesiredInterception(null);
        }
    }

    private void updateGesture(BackEventCompat event) {
        lastProgress = clamp(event.getProgress());
        lastTouchX = event.getTouchX();
        lastTouchY = event.getTouchY();
        switch (event.getSwipeEdge()) {
            case BackEventCompat.EDGE_LEFT:
                lastEdge = EDGE_LEFT;
                break;
            case BackEventCompat.EDGE_RIGHT:
                lastEdge = EDGE_RIGHT;
                break;
            default:
                lastEdge = EDGE_NONE;
                break;
        }
    }

    private void emitCurrent(@NonNull String phase, @NonNull String source) {
        emit(phase, lastProgress, source, lastEdge, lastTouchX, lastTouchY);
    }

    private void resetGesture() {
        gestureStarted = false;
        gestureFinishing = false;
        gestureTarget = null;
        gestureInterceptorId = "";
        gestureId = 0;
        lastProgress = 0;
        lastTouchX = 0;
        lastTouchY = 0;
        lastEdge = EDGE_NONE;
    }

    private void applyDesiredInterception(@Nullable Callback callback) {
        if (destroyed) {
            if (callback != null) {
                callback.invoke("Back has already been destroyed");
            }
            return;
        }
        FragmentActivity host = activity;
        if (host == null || host.isFinishing() || host.isDestroyed()) {
            if (callback != null) {
                callback.invoke("Back requires a usable FragmentActivity host");
            }
            return;
        }
        if (!registered) {
            host.getOnBackPressedDispatcher().addCallback(host, backCallback);
            registered = true;
        }
        backCallback.setEnabled(desiredEnabled);
        if (callback != null) {
            callback.invoke("");
        }
    }

    private void emit(
            String phase,
            double progress,
            String source,
            String edge,
            double touchX,
            double touchY) {
        LynxView lynxView = mLynxContext == null ? null : mLynxContext.getLynxView();
        if (lynxView == null) {
            return;
        }
        JavaOnlyMap payload = new JavaOnlyMap();
        payload.putString("platform", PLATFORM_ANDROID);
        payload.putString("phase", phase);
        payload.putDouble("progress", clamp(progress));
        payload.putString("source", source);
        payload.putString("edge", edge);
        payload.putDouble("touchX", touchX);
        payload.putDouble("touchY", touchY);
        payload.putString("interceptorId", gestureInterceptorId);
        payload.putDouble("gestureId", gestureId);
        lynxView.sendGlobalEvent(EVENT_NAME, JavaOnlyArray.of(payload));
    }

    private static double clamp(double value) {
        return Math.max(0, Math.min(value, 1));
    }

    @Nullable
    private static FragmentActivity resolveFragmentActivity(@Nullable Context context) {
        Context current = context;
        while (current instanceof ContextWrapper) {
            if (current instanceof FragmentActivity) {
                return (FragmentActivity) current;
            }
            current = ((ContextWrapper) current).getBaseContext();
        }
        return current instanceof FragmentActivity ? (FragmentActivity) current : null;
    }
}

File: app/androidApp/app/src/main/java/com/lynxapp/nativemodule/BackModule.kt
package com.lynxapp.nativemodule

import android.app.Activity
import android.os.Build
import android.window.BackEvent
import android.window.OnBackAnimationCallback
import android.window.OnBackInvokedCallback
import android.window.OnBackInvokedDispatcher
import androidx.annotation.RequiresApi
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.tasm.LynxView

/**
 * Activity-owned back dispatcher shared with the page's BackModule.
 *
 * The enabled bit is updated before a system gesture begins. This keeps the
 * platform decision synchronous while JS receives lifecycle events later.
 */
class NativeBackController(private val activity: Activity) {
    private var lynxView: LynxView? = null
    private var enabled = false
    private var platformCallback: Any? = null
    private var gestureStarted = false
    private var lastProgress = 0.0
    private var lastTouchX = 0.0
    private var lastTouchY = 0.0
    private var lastEdge = EDGE_NONE

    fun attach(lynxView: LynxView) {
        this.lynxView = lynxView
    }

    fun setEnabled(nextEnabled: Boolean) {
        if (enabled == nextEnabled) {
            return
        }
        if (!nextEnabled && gestureStarted) {
            emit(
                phase = PHASE_CANCEL,
                progress = lastProgress,
                source = SOURCE_GESTURE,
                edge = lastEdge,
                touchX = lastTouchX,
                touchY = lastTouchY,
            )
            resetGesture()
        }
        enabled = nextEnabled
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (nextEnabled) {
                registerPlatformCallback()
            } else {
                unregisterPlatformCallback()
            }
        }
    }

    /** Called by Activity.onBackPressed for Android 12L and earlier. */
    fun handleLegacyBack(): Boolean {
        if (!enabled) {
            return false
        }
        emitDiscreteBack(SOURCE_SYSTEM)
        return true
    }

    fun destroy() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            unregisterPlatformCallback()
        }
        enabled = false
        gestureStarted = false
        lynxView = null
    }

    private fun registerPlatformCallback() {
        if (platformCallback != null) {
            return
        }
        platformCallback = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            Api34.createAndRegister(activity, this)
        } else {
            Api33.createAndRegister(activity, this)
        }
    }

    private fun unregisterPlatformCallback() {
        val callback = platformCallback ?: return
        Api33.unregister(activity, callback)
        platformCallback = null
        resetGesture()
    }

    private fun emitDiscreteBack(source: String) {
        emit(
            phase = PHASE_START,
            progress = 0.0,
            source = source,
            edge = EDGE_NONE,
            touchX = 0.0,
            touchY = 0.0,
        )
        emit(
            phase = PHASE_COMMIT,
            progress = 1.0,
            source = source,
            edge = EDGE_NONE,
            touchX = 0.0,
            touchY = 0.0,
        )
    }

    private fun onBackStarted(event: BackEvent) {
        gestureStarted = true
        updateGesture(event)
        emitCurrent(PHASE_START)
    }

    private fun onBackProgressed(event: BackEvent) {
        if (!gestureStarted) {
            gestureStarted = true
            updateGesture(event)
            emitCurrent(PHASE_START)
        } else {
            updateGesture(event)
        }
        emitCurrent(PHASE_PROGRESS)
    }

    private fun onBackCancelled() {
        if (!gestureStarted) {
            return
        }
        emitCurrent(PHASE_CANCEL)
        resetGesture()
    }

    private fun onBackInvoked() {
        if (gestureStarted) {
            emit(
                phase = PHASE_COMMIT,
                progress = 1.0,
                source = SOURCE_GESTURE,
                edge = lastEdge,
                touchX = lastTouchX,
                touchY = lastTouchY,
            )
            resetGesture()
        } else {
            emitDiscreteBack(SOURCE_SYSTEM)
        }
    }

    private fun updateGesture(event: BackEvent) {
        lastProgress = event.progress.toDouble().coerceIn(0.0, 1.0)
        lastTouchX = event.touchX.toDouble()
        lastTouchY = event.touchY.toDouble()
        lastEdge = when (event.swipeEdge) {
            BackEvent.EDGE_LEFT -> EDGE_LEFT
            BackEvent.EDGE_RIGHT -> EDGE_RIGHT
            else -> EDGE_NONE
        }
    }

    private fun emitCurrent(phase: String) {
        emit(
            phase = phase,
            progress = lastProgress,
            source = SOURCE_GESTURE,
            edge = lastEdge,
            touchX = lastTouchX,
            touchY = lastTouchY,
        )
    }

    private fun resetGesture() {
        gestureStarted = false
        lastProgress = 0.0
        lastTouchX = 0.0
        lastTouchY = 0.0
        lastEdge = EDGE_NONE
    }

    private fun emit(
        phase: String,
        progress: Double,
        source: String,
        edge: String,
        touchX: Double,
        touchY: Double,
    ) {
        val payload = JavaOnlyMap().apply {
            putString("platform", PLATFORM_ANDROID)
            putString("phase", phase)
            putDouble("progress", progress.coerceIn(0.0, 1.0))
            putString("source", source)
            putString("edge", edge)
            putDouble("touchX", touchX)
            putDouble("touchY", touchY)
        }
        lynxView?.sendGlobalEvent(EVENT_NAME, JavaOnlyArray.of(payload))
    }

    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    private object Api33 {
        fun createAndRegister(
            activity: Activity,
            controller: NativeBackController,
        ): OnBackInvokedCallback {
            val callback = OnBackInvokedCallback(controller::onBackInvoked)
            activity.onBackInvokedDispatcher.registerOnBackInvokedCallback(
                OnBackInvokedDispatcher.PRIORITY_DEFAULT,
                callback,
            )
            return callback
        }

        fun unregister(activity: Activity, callback: Any) {
            activity.onBackInvokedDispatcher.unregisterOnBackInvokedCallback(
                callback as OnBackInvokedCallback,
            )
        }
    }

    @RequiresApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
    private object Api34 {
        fun createAndRegister(
            activity: Activity,
            controller: NativeBackController,
        ): OnBackAnimationCallback {
            val callback = object : OnBackAnimationCallback {
                override fun onBackStarted(backEvent: BackEvent) {
                    controller.onBackStarted(backEvent)
                }

                override fun onBackProgressed(backEvent: BackEvent) {
                    controller.onBackProgressed(backEvent)
                }

                override fun onBackCancelled() {
                    controller.onBackCancelled()
                }

                override fun onBackInvoked() {
                    controller.onBackInvoked()
                }
            }
            activity.onBackInvokedDispatcher.registerOnBackInvokedCallback(
                OnBackInvokedDispatcher.PRIORITY_DEFAULT,
                callback,
            )
            return callback
        }
    }

    companion object {
        const val EVENT_NAME = "back"
        private const val PLATFORM_ANDROID = "android"
        private const val PHASE_START = "start"
        private const val PHASE_PROGRESS = "progress"
        private const val PHASE_CANCEL = "cancel"
        private const val PHASE_COMMIT = "commit"
        private const val SOURCE_SYSTEM = "system"
        private const val SOURCE_GESTURE = "gesture"
        private const val EDGE_LEFT = "left"
        private const val EDGE_RIGHT = "right"
        private const val EDGE_NONE = "none"
    }
}

/** Lets the current Lynx page synchronously opt in or out of native back interception. */
class BackModule(context: android.content.Context, param: Any?) : LynxModule(context, param) {
    private val activity = findActivity(context)
    private val controller = param as? NativeBackController

    private fun findActivity(context: android.content.Context?): Activity? {
        var current = context
        while (current is android.content.ContextWrapper) {
            if (current is Activity) return current
            current = current.baseContext
        }
        return null
    }

    @LynxMethod
    fun setEnabled(enabled: Boolean, callback: Callback) {
        val host = activity
        val backController = controller
        if (host == null || backController == null) {
            callback.invoke("Back has no Activity host")
            return
        }
        host.runOnUiThread {
            backController.setEnabled(enabled)
            callback.invoke("")
        }
    }

    override fun destroy() {
        activity?.runOnUiThread {
            controller?.setEnabled(false)
        }
        super.destroy()
    }

    companion object {
        const val NAME = "Back"
    }
}
